File: 0036-valid-sudoku/0036-valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {

        unordered_map<int, unordered_set<int>> rows;
        unordered_map<int, unordered_set<int>> columns;
        unordered_map<string, unordered_set<int>> square;

        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (board[i][j] == '.') {
                    continue;
                }

                string key = to_string(i/3) + ", " + to_string(j/3);

                if (rows[i].count(board[i][j]) || columns[j].count(board[i][j]) ||
                square[key].count(board[i][j])) {
                    return false;
                }

                rows[i].insert(board[i][j]);
                columns[j].insert(board[i][j]);
                square[key].insert(board[i][j]);
            }
        }

        return true;


        
    }
};

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = [set() for _ in range(9)]
        column = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]

        for i in range(0, 9):
            for j in range(0, 9):
                val = board[i][j]
                if val == ".":
                    continue
                
                if val in row[i]:
                    return False
                
                if val in column[j]:
                    return False

                box = (i // 3) * 3 + (j // 3)

                if val in boxes[box]:
                    return False
                
                row[i].add(val)
                column[j].add(val)
                boxes[box].add(val)
            
        return True
                
        